In [98]:
# Experiment 9
#Aim - A. To generate, read and write audio signals
#      B. To compress audio signal
#Name - Param Nagda
#Roll No - C035
#Class - BTech EXTC
#Batch - B2
#Date of Experiment - 16/09/20
#Date of Submission - 19/09/20

In [1]:
import numpy as np
import librosa as lr
import IPython.display as ipd
import soundfile as sf
import matplotlib.pyplot as plt
from scipy.fftpack import dct,idct
from sklearn.metrics import mean_squared_error

In [2]:
x, sr = lr.load('marySongOrig.wav')

In [3]:
ipd.Audio(x, rate=sr)

In [4]:
sf.write('new_mary.wav',x,sr)

In [5]:
y, sr2 = lr.load('new_mary.wav')

In [6]:
sr2 = 15000
ipd.Audio(y, rate=sr2)

In [7]:
def compress_audio(x,t):   #img:Image, t:Threshold Percentage
    N=8
    m = x.shape[0]
    
    # DCT of image
    audiodct = dct(x,axis=0,norm='ortho')
    #Compressiion Threshold
    th = np.sort(audiodct)[::-1][int((m-1)*t)]

    
    # Compression in DCT
    audiocdct = np.zeros(m, dtype=int)
    for i in range(m):
            if np.abs(audiodct[i]) > th:
                audiocdct[i] = audiodct[i]
            else:
                audiocdct[i] = 0

                
    # IDCT
    audionew = idct(audiocdct,axis=0,norm='ortho')
            
    # Compression Ratio
    nonzeros = np.sum([audiocdct!=0])
    
    return (audionew,round(m/nonzeros,2))

In [92]:
z, cr = compress_audio(x,0.01)

In [93]:
ipd.Audio(z, rate=sr)

In [94]:
z

array([0.03529206, 0.03466681, 0.03342781, ..., 0.01952952, 0.02060794,
       0.02115375])

In [95]:
cr

668.88

In [96]:
audio_mse = mean_squared_error(x,z)

In [97]:
audio_mse

0.0019350633531167046

### Conclusion:-

Sound is a physical disturbance in a medium. It propagates in the medium as pressure wave by the movement of atoms or molecules. Here using the librosa library we were able to read and write into an audio file. We even changed the rate of the audio and observed the difference. We then applied DCT and IDCT to compress and transmit the audio signal. By applying various threshold levels we observed that at 0.01% the audio has alot of distortion with CR (Compression Ratio) of 4995.72, increasing the threshold to 0.05 the audio was comparatively less distorted while the CR decreased to 1024.76. At threshold level of 1% or more the audio the almost similar to the original one and the CR being constant at 668.88. The MSE (Mean Square Error) between the original audio and the compressed audio observed at threshold level of 0.01%,0.05%,1%+ were 0.0055,0.0027,0.0019 respectively.